packages/flutter/lib/src/rendering/animated_size.dart

```diff
@@ -0,0 +1,100 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/animation.dart';
+
+import 'box.dart';
+import 'object.dart';
+import 'shifted_box.dart';
+
+/// A render object that can render child size change animations.
+class RenderAnimatedSize extends RenderAligningShiftedBox {
+  /// Creates a render animated size.
+  RenderAnimatedSize({
+    Duration duration,
+    Curve curve,
+    RenderBox child
+  }) : super(child: child, alignment: FractionalOffset.center) {
+    _controller = new AnimationController(
+      duration: duration
+    );
+    _animation = new CurvedAnimation(
+      parent: _controller,
+      curve: curve
+    );
+  }
+
+  AnimationController _controller;
+  CurvedAnimation _animation;
+  Size _oldSize;
+  Size _newSize;
+
+  /// The duration of the animation.
+  Duration get duration => _controller.duration;
+  set duration(Duration value) {
+    assert(value != null);
+    if (value == _controller.duration)
+      return;
+    _controller.duration = value;
+  }
+
+  /// The curve of the animation.
+  Curve get curve => _animation.curve;
+  set curve(Curve value) {
+    assert(value != null);
+    if (value == _animation.curve)
+      return;
+    _animation.curve = value;
+  }
+
+  @override
+  void detach() {
+    _controller?.stop();
+    super.detach();
+  }
+
+  void _trigger() {
+    _controller
+      ..removeListener(markNeedsLayout)
+      ..value = 0.0
+      ..forward()
+      ..addListener(markNeedsLayout);
+  }
+
+  Size _getSize() {
+    return Size.lerp(_oldSize, _newSize, _animation.value);
+  }
+
+  @override
+  void performLayout() {
+    if (child != null) {
+      child.layout(constraints, parentUsesSize: true);
+
+      if (_newSize == null) {
+        _oldSize = _newSize = constraints.constrain(child.size);
+      } else {
+        final Size childSize = constraints.constrain(child.size);
+
+        if (_newSize != childSize) {
+          _oldSize = _newSize;
```

Comment:
This will cause a visual jump in the following situation:

1) Layout at a given size.
2) Layout a a new size.
3) During the animation to the new size, layout at a third size.

At this point, the size will jump to the end of the first animation instead of picking at the current size.

To solve this problem, you need to remember three sizes:

1) `_sourceSize`
2) `_targetSize`
3) `_lastSize`

Where `_lastSize` is the size the child was the most recent time through layout.  When changing the `_targetSize`, you need to set the `_sourceSize` to the `_lastSize`.

When you do this you need to be careful because you set the animation to 0.0, which means you'll produce two frames with the same size (you'll repeat `_lastSize`).  Instead of resetting the animation to 0.0, if the animation is already running, you need to remember the `_lastValue`, which is the value the animation had the last time through layout.  You need to start the new animation at `animation.value - _lastValue`.  That way the frame you produce now will be different from the frame you produced last time.  It will be as if you when back in time and knew the animation was going to advance already when you produced the last frame.

/cc @Hixie